clippy_lints/src/extra_unused_type_parameters.rs

```diff
@@ -151,11 +175,15 @@ impl<'cx, 'tcx> Visitor<'tcx> for TypeWalker<'cx, 'tcx> {
         if let WherePredicate::BoundPredicate(predicate) = predicate {
             // Collect spans for any bounds on type parameters. We only keep bounds that appear in
             // the list of generics (not in a where-clause).
-            //
-            // Also, if the function body is empty, we don't lint the corresponding type parameters
-            // (See https://github.com/rust-lang/rust-clippy/issues/10319).
             if let Some((def_id, _)) = predicate.bounded_ty.peel_refs().as_generic_param() {
-                if self.fn_body_empty {
+                // If the bound contains non-public traits, err on the safe side and don't lint the
+                // corresponding parameter.
+                if !predicate
+                    .bounds
+                    .iter()
+                    .filter_map(bound_to_trait_def_id)
+                    .all(|def_id| self.cx.effective_visibilities.is_exported(def_id))
+                {
                     self.ty_params.remove(&def_id);
```

Comment:
Should also set `all_params_unused`.

Reply:
Good catch - this was actually introduced in #10321.